cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,225 @@
+# CFP-27752: Operator Manages Cilium Identities
+
+**Authors:** dordel@google.com
+
+**Area:** SIG: SIG-POLICY, SIG-AGENT, SIG-K8S
+
+**Target Cilium Release**: 1.17
+
+**Status**: Implementable
+
+**Sharing:** Public
+
+**Updated:** 2024-09-13
+
+**Reviewers:** [Antonio Ojea](https://github.com/aojea), [Joe Stringer](https://github.com/joestringer)
+
+Github issue: [https://github.com/cilium/cilium/issues/27752](https://github.com/cilium/cilium/issues/27752)
+
+Agent and Operator are short names used for cilium-agent and cilium-operator respectively.
+
+Abbreviations for Cilium custom resources used in this document:
+
+**CID  - [Cilium Identity](https://docs.cilium.io/en/stable/gettingstarted/terminology/#identity)**
+
+**CEP - [Cilium Endpoint](https://docs.cilium.io/en/latest/network/kubernetes/ciliumendpoint/)**
+
+**CES - [Cilium Endpoint Slice](https://docs.cilium.io/en/latest/network/kubernetes/ciliumendpointslice/)**
+
+# Summary
+
+Centralize CID management to a single pod (Operator), instead of distributed management, done by all Agents. The goal is to improve security, reliability, performance and scalability of CID, and enable more advanced optimizations.
+
+# Background & Motivation
+
+CID represents complete data of one security identity, consisting of a numeric identity and a list of labels. Each CEP has an assigned identity based on its pod and namespace labels. CIDs are used to group pod and namespace labels that are used to enforce network policies.
+
+**CID duplication** is a flaw of the CID management system. It occurs whenever Agents create multiple CIDs for the same set of labels. The most common case is when pods of a new deployment (new unique label set) is scheduled onto multiple nodes. Although deployment creation causes CID duplication, it doesn’t happen often enough to produce many duplicate CIDs. However, namespace label changes, either on pods or namespaces, immediately trigger massive CID duplication, because all Agents react to the changes at the same time. A test with 5000 nodes shows that having 1 deployment with 5000 pods running 1 pod per node, can cause 4999 duplicate CIDs after changing the namespace labels.
+
+CIDs are presenting issues with reliability and scalability, and are causing network policies to be in a broken state. Most notable cases:
+
+* Reaching maximum number (65k) of CIDs in the cluster.
+* eBPF policy maps overflowing with over 16k entries.
+
+# Design
+
+## Operator changes
+
+Operator will start managing CIDs, and will become the only system pod that issues mutating requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing (creating) CIDs.
```

Comment:
```suggestion
Operator will start managing CIDs, and will become the only system pod that issues creation or deletion requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing (creating) CIDs.
```

I mean to clarify here that these resources are intended to be immutable and never to be updated; mutation in this context is ambiguous.

Reply:
(I see that later text reflects this immutable understanding and even implements safeguards to undo externally-sourced mutations)